### PR TITLE
Add on-target tests to CI

### PR DESCRIPTION
**Pull request by newAM**
A week ago in the rust-embedded matrix chat @adamgreig mentioned that on-target CI would be helpful for the `cortex-m` and `cortex-m-rt` crates.  This is a pull-request to make that happen.

## History: Bootstrapping and `cortex-m-rt`

The first problem I came across was the amount of boilerplate required to build an on-target binary without `cortex-m-rt`.

There are two paths forward here:

1. Introduce a lot of boilerplate in `cortex-m`, which will largely be copy-pasted from `cortex-m-rt`.
2. Relocate `cortex-m-rt` to this workspace.

In (#391) `cortex-m-rt` was relocated to this workspace to fix the bootstrapping problem.

## Test Execution

Tests run with QEMU and physical hardware.

QEMU uses the LM3S6965 Cortex-M3 target because it is widely used.

The physical hardware is a NUCLEO-F070RB connected to a self-hosted runner.  In the future more hardware can be added to cover more CPUs.

Due to reliability concerns with self-hosted runners only QEMU will be a required status check in CI, the physical hardware checks will be informational only.

### CI Software

The CI software for running on physical hardware is simply a self-hosted github actions runner.  A working demonstration of this can be found [here](https://github.com/newAM/totally-not-a-cortex-m-fork/runs/5345451343?check_suite_focus=true) (the repository does not appear as a fork to work around github actions limitations with forks).

The runner uses [`probe-run`] to execute the tests on embedded hardware.  [`probe-run`] was chosen for several reasons:

* Actively maintained by [knurling-rs], with an actively maintained back-end from [`probe-rs`].
* Written in rust.  Understanding the code does not require contributors to learn a new language.
* Designed with on-target testing as a primary goal (for use with [`defmt-test`]).
* Automatic unwinding and backtrace display.

## Test Harness

This PR introduces a test harness, `minitest`.

`minitest` is almost identical to [`defmt-test`], the only difference is that it replaces [`defmt`] with [`rtt-target`] because [`defmt`] introduces a dependency cycle on `cortex-m`.

This is harness is very minimal, adding only 327 lines of rust:

```console
$ tokei testsuite/minitest 
===============================================================================
 Language            Files        Lines         Code     Comments       Blanks
===============================================================================
 Markdown                1            7            0            4            3
 TOML                    2           41           34            0            7
-------------------------------------------------------------------------------
 Rust                    3          406          350            5           51
 |- Markdown             1            8            0            7            1
 (Total)                            414          350           12           52
===============================================================================
 Total                   6          454          384            9           61
===============================================================================
```

The test harness does introduce some abstraction, and may not be suitable for all tests.  Lower-level tests are still possible without the harness using `asm::udf` to fail the test, and `asm::bkpt` to exit without failure.

## Reliability and Uptime

I have been doing automatic on-target testing for the [`stm32wlxx-hal`] using [`probe-run`].  Over hundreds of automatic runs spanning several months I have had no failures as a result of external factors (USB connectivity, programming errors, ect.).  I do not anticipate on-target CI being perfect, but at the same time I do not anticipate frequent problems.

[`defmt-test`]: https://github.com/knurling-rs/defmt/tree/main/firmware/defmt-test
[`defmt`]: https://ferrous-systems.com/blog/defmt/
[`probe-rs`]: https://probe.rs/
[`probe-run`]: https://github.com/knurling-rs/probe-run
[`rtt-target`]: https://crates.io/crates/rtt-target
[`stm32wlxx-hal`]: https://github.com/stm32-rs/stm32wlxx-hal
[knurling-rs]: https://knurling.ferrous-systems.com/

### CODE REVIEW

**Comment by newAM** [comment]
This was discussed in the weekly meeting: https://github.com/rust-embedded/wg/blob/master/minutes/2021-10-26.md

There is some hesitation to add on-target testing without an SLA, which is completely understandable.  Based on that feedback I updated the draft PR:

* Changed CI to use QEMU instead of the self-hosted runner
* Added compile-time features to select RTT or semihosting
  * Allows for the flexibility to easily run on-target or run with QEMU
* Moved `memory.x` to the `testsuite` directory per @adamgreig 's feedback

The tests can still be run on physical hardware.  There is still a discussion to be resolved as to what approach to take with regards to gating/non-gating and automatic/manual.

@thejpster noted that the Cortex-M55 FPV is free for use (no licensing issues).  I plan to explore that option this weekend as well.

**Comment by newAM** [comment]
I updated this pull-request for #391 which solves the bootstrapping problem; CI is passing now! :tada:

I also removed the demonstration do-nothing tests and added a couple basic tests.

**Comment by adamgreig** [comment]
This is looking great, thanks! Perhaps it would be worth adding the new CI flow to bors' required statuses?

I'd still be interested in having it run automatically on hardware as a best-effort thing: perhaps qemu on GitHub's own CI and required for bors, and also (tries to) runs a job on your hardware, but without being required for bors so it wouldn't block merging if it wasn't working?

Is there anything else before this is "ready for review"?

**Comment by newAM** [comment]
> This is looking great, thanks! Perhaps it would be worth adding the new CI flow to bors' required statuses?

Missed that, fixed, thanks!

> I'd still be interested in having it run automatically on hardware as a best-effort thing: perhaps qemu on GitHub's own CI and required for bors, and also (tries to) runs a job on your hardware, but without being required for bors so it wouldn't block merging if it wasn't working?
> 
> Edit: To be clear, no need for on-hardware testing as part of this PR or anything, we could go ahead with just qemu if you like. Just making it clear I still think there's a lot of value in automated on-hardware testing too, in addition to it being easy to run on-hardware tests manually from the repo.

I can do that, everything is setup because I run on-target testing for the `stm32wlxx-hal` anyway.

@thejpster expressed some SLA concerns before - what do you think about this approach instead?

> Is there anything else before this is "ready for review"?

Just cleaning up the introduction message so that the now-irrelevant details do not get committed by bors :smile:

**Comment by newAM** [comment]
@thejpster any concerns with the above - adding on-target testing with real hardware to CI (in addition to QEMU), but without the real hardware CI runs being a gating check to mitigate SLA issues?

**Review by thejpster — DISMISSED** [review]
Looks great. Love the really thorough documentation.

**Comment by thejpster** [comment]
Apologies it took so long to get to this, too.

**Comment by newAM** [comment]
> Apologies it took so long to get to this, too.

No problem!

> > Is there anything else before this is "ready for review"?
> 
> Just cleaning up the introduction message so that the now-irrelevant details do not get committed by bors smile

I cleaned up the PR body to reflect the current plan, re-added the HIL tests, and marked this as ready for review :+1:

Before this gets merged the self-hosted runner needs to get added; I will need the github actions runner registration token for this repository, or access to the repository to do this myself.

**Comment by newAM** [comment]
The final steps are done and this should be good to go now!  I added the self-hosted runner and rebased onto `origin/master`.

Repo admins can see the runner in the settings here: https://github.com/rust-embedded/cortex-m/settings/actions/runners

Here's the new job with the runner: https://github.com/rust-embedded/cortex-m/actions/runs/2048592598
The `hil-compile-rtt` job runs on the github runners, then it passes the binaries off to `hil-stm32` to test them on the self-hosted runner.

**Review by adamgreig — APPROVED** [review]
LGTM, thanks for your perseverance! I'll leave this to give other @rust-embedded/cortex-m peolpe a chance to look over and if there's no objections we can merge in a couple of days?

**Review by thejpster — APPROVED** [review]

**Pull request merged**
